Remove `@progress_step` decorator from ZHA and Hardware integration

puddly opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This is a proper fix for the issue uncovered by #153906 and migrates ZHA and the hardware integration firmware installation flows to use the more verbose API.

This fixes an issue where the ZHA abort dialog did not show in some cases when network settings could not be written to the new coordinator.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.

  AI tools are welcome, but contributors are responsible for *fully*
  understanding the code before submitting a PR.
-->

- [x] I understand the code I am submitting and can explain how it works.
- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.
- [x] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

puddly committed:
Get rid of `@progress_step()` decorator usage in ZHA

puddly committed:
Get rid of `@progress_step()` usage in Hardware firmware config flows

puddly committed:
Fix failing test

puddly committed:
Fix final failing progress error, in ZHA

puddly committed:
Fix rebase issues

puddly committed:
Avoid having tasks create steps, keep step logic in progress step

MartinHjelmare commented on homeassistant/components/zha/config_flow.py:
```diff
@@ -735,10 +788,11 @@ async def async_step_restore_backup(
             return self.async_show_progress_done(next_step_id="pre_plug_in_new_radio")
         except CannotWriteNetworkSettings as exc:
-            return self.async_abort(
+            self._progress_error = AbortFlow(
```
If this didn't fail any tests before, it means we're lacking tests for this case.

TheJulianJES replied:
Yep, kind of. The test for this currently doesn't raise the "invalid transition" ValueError, because the progress flow never really starts, as we instantly raise `CannotWriteNetworkSettings` in the test. 

Just adding `asyncio.sleep(0)` to the `side_effect` before raising `CannotWriteNetworkSettings` causes the test to fail as expected, without this PR.
So, we should add something along those lines to the test to make sure the progress flow has a chance to start.

For reference, the test for this is:
https://github.com/home-assistant/core/blob/c0db966afd1b5b8ddf3689ad9dec962cfe5ef525/tests/components/zha/test_config_flow.py#L721-L739

puddly committed:
Use `DelayedAsyncMock` to ensure progress steps are well tested

puddly committed:
Fix tests requiring `consume_progress_flow`

TheJulianJES reviewed (changes requested):
### ZHA

I tested all three ZHA progress flows again. Everything still works fine there.

I can also confirm that the `AbortFlow` for `CannotWriteNetworkSettings` now works properly, even when it's not immediately raised before the progress flow has a chance to start. In both cases, we now correctly transition to the abort step and show the abort reason.


### HA hardware integrations

The Home Assistant hardware integrations abort dialogs all no longer work correctly, as no description placeholders are provided now.
This was previously done in the `_failed` steps, but they're no longer used at all, as `AbortFlow`s are now raised more directly (using the intermediate progress failure step), but without any description placeholders.

You can verify this by removing these three methods: `async_step_fw_download_failed`, `async_step_firmware_install_failed`, and `async_step_unsupported_firmware`. Tests will still pass!

There are two main soloutions I can see here:

1. Go back to properly using the individual failure steps that provide the description placeholders.
  (If we do that, we should also rename the `async_step_fw_download_failed` back to `async_step_firmware_download_failed`, since the other two install + unsupported firmware failure steps also do not shorten `firmware` to `fw`. This also adds to a bit of confusion with the similarly named translation keys otherwise.)
2. Remove the three (unused) failure steps entirely and just provide the relevant description placeholders, where we now raise `AbortFlow`s that mostly directly propagate to the frontend.
  This would need to be done in quite a few places though.
3. Mix of both: use normal methods to get/raise an `AbortFlow` with proper description placeholders for each of the possible common failure cases, like `firmware_download_failed`.
  This should avoid some duplicate code of needing to set the description placeholders everywhere where `firmware_download_failed` can be raised, for example.
  Still, this would be closer to the pattern used in the ZHA config flow and avoids the (IMO) unintuitive use of an `AbortFlow` `reason` for the next step, which only then raises an `AbortFlow` with proper description placeholders used by the frontend...

I think I like option 3 of those the most.

TheJulianJES commented on homeassistant/components/homeassistant_hardware/firmware_config_flow.py:
```diff
@@ -192,9 +198,8 @@ async def _install_firmware_step(
-                next_step_id=err.reason,
-            )
+            self._progress_error = err
+            return self.async_show_progress_done(next_step_id="progress_failed")
```
Example:
<img width="572" height="214" alt="image" src="https://github.com/user-attachments/assets/eddc81cd-d168-4807-8485-b29e1064e661" />

TheJulianJES replied:
Confirmed fixed with [`d6b4f88` (#155867)](https://github.com/home-assistant/core/pull/155867/commits/d6b4f88ed3d0235d13d43d44670097779d901b60)

puddly committed:
Always raise a fully translated `AbortFlow`

TheJulianJES commented on homeassistant/components/homeassistant_hardware/firmware_config_flow.py:
nit/note: Since we always add `"firmware_name"` in `_get_translation_placeholders` now, we technically don't need to explicitly add the `"firmware_name"` here too.

It won't cause any issues if we leave it, but the `_install_firmware` method (executed in task) sets `self.installing_firmware_name` to the same firmware name directly at the beginning of the method, so `self._get_translation_placeholders()` will correctly return the firmware name.

As the task is started eagerly, it's always set in time.

TheJulianJES replied:
A more critical issue here is that we do not test the `"firmware_name"` description placeholder in the (non-failure) config flow tests:
https://github.com/home-assistant/core/blob/d52749c71ae64fc953b7f342565a6aaf89f10ef2/tests/components/homeassistant_hardware/test_config_flow.py#L796-L798

It might be nice to assert all provided description placeholders in the test at once, same as done in the config flow failure tests now, even if not all description placeholders are used by all strings (but that was already the case before with some of the abort dialogs).

TheJulianJES reviewed (commented):
Looks good to me. Also manually tested the errors in the hardware integration again and they work properly now.
One small note that doesn't cause any issues though:

TheJulianJES committed:
Remove redundant `firmware_name` description placeholder

This is already included by `self._get_translation_placeholders()`, so no need to replace it with the same value.

TheJulianJES committed:
Assert all description placeholders in tests

TheJulianJES reviewed (approved):
Tested all ZHA and hardware integration steps as well. All still working as expected.

This PR does fix an issue in ZHA where the abort dialog, when we fail writing network settings (`CannotWriteNetworkSettings`), may not be shown in some cases.
It's a rare case to hit in the first place, but I'm tagging the PR for the patch release today for that reason.

Otherwise, we also don't use the buggy `@progress_step` decorator anymore and the tests are much improved, addressing Martin's comment from here: https://github.com/home-assistant/core/pull/155867#discussion_r2498242838
